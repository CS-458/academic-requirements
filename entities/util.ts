let id = 0;
export function uniqueId(): string {
  return `id_${id++}`;
}
