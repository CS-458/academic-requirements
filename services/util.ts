/// Fetch a request (similar to fetch), but throw an error if the request fails
export async function fetchApi<T>(url: string, init?: RequestInit): Promise<T> {
  const res = await fetch(url, init);
  if (!res.ok) {
    throw new Error(`Response code ${res.status} for request to ${url}`);
  }
  return await res.json();
}
