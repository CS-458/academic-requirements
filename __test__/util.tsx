import { jest } from "@jest/globals";
import type { MatcherFunction } from "expect";
import { DndProvider } from "react-dnd";
import { HTML5Backend } from "react-dnd-html5-backend";
import { QueryClient, QueryClientProvider, UseQueryResult } from "react-query";
import { testApiHandler } from "next-test-api-route-handler";
import { NextApiHandler } from "next";
import {
  Matcher,
  screen,
  render as testRender,
  RenderResult,
  waitFor
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { UserEvent } from "@testing-library/user-event/dist/types/setup/setup";
import { User, UserMajor } from "../services/user";
import { VerifyIdTokenOptions, LoginTicket } from "google-auth-library";
import { updateClient } from "../services/login";
import { PromisedDatabase } from "promised-sqlite3";
import { setUserDb } from "../services/sql";

import "../pages/api";
import { dragAndDrop } from "./dragDrop";

interface UserExt {
  /// Select an element from an `AutoComplete` dropdown
  selectAutocomplete: typeof selectAutocomplete;
  /// Drag an element from one place to another
  drag: typeof dragAndDrop;
}

async function selectAutocomplete(
  this: UserEvent,
  label: Matcher,
  option: Matcher
): Promise<void> {
  await this.click(screen.getByTestId("test-root-element"));
  const input = screen.getByLabelText(label, { selector: "input" });
  await waitFor(async () => {
    expect(input).not.toBeDisabled();
  });
  await this.click(input);
  try {
    await this.click(
      await screen.findByText(option, undefined, { timeout: 10000 })
    );
  } catch (e) {
    const rootEl = screen.getByTestId("test-root-element");
    const popper = rootEl?.querySelector(".MuiAutocomplete-popper");
    const html = popper?.innerHTML;
    throw new Error(
      `Option ${option} was not found in dropdown.\nOptions: ${html}`
    );
  }
}

export function parentEl(el: HTMLElement, name: string): HTMLElement {
  while (!el.classList.contains(`${name}-root`)) {
    const next = el.parentElement;
    if (next == null) {
      throw new Error(`Element does not have a wrapper with name ${name}`);
    }
    el = next;
  }
  return el;
}

/// userEvent.setup(), but with some additional convience methods
export function setupUser(): UserEvent & UserExt {
  return {
    ...userEvent.setup(),
    selectAutocomplete,
    drag: dragAndDrop
  };
}

// const modules: { [key: string]: NextApiHandler | Promise<NextApiHandler> } = {};
// declare function fetch2(input: RequestInfo | URL, init?: RequestInit): Promise<Response>;
/// Executes a request against an Api Route, roughly equavelent with `fetch`.
export async function fetchApiRoute(
  url: RequestInfo | URL,
  init?: RequestInit
): Promise<Response> {
  const res = { resolve: (_a: any) => { }, reject: (_e: any) => { } };
  const result: Promise<Response> = new Promise((resolve, reject) => {
    res.resolve = resolve;
    res.reject = reject;
  });

  let fullUrl: string | undefined;
  let path;
  let params;
  if (url instanceof URL) {
    fullUrl = url.href;
    path = url.pathname;
    params = url.search;
  } else if (typeof url === "string") {
    fullUrl = url;
    [path, params] = url.split("?");
  } else {
    fullUrl = url.url;
    [path, params] = url.url.split("?");
  }

  const handler: NextApiHandler = await import(`../pages/${path}`).catch(
    (e) => {
      if (jest.isEnvironmentTornDown()) {
        console.log("Request made after environment torn down");
      } else {
        console.error(e);
      }
    }
  );
  await testApiHandler({
    url: fullUrl,
    handler,
    params: params?.split("&").reduce((a, b) => {
      const [name, value] = b.split("=");
      return { ...a, [name]: value };
    }, {}),
    test: async ({ fetch }) => {
      const inner = await fetch(init);
      res.resolve(inner);
    }
  });
  return await result;
}

/// Calls fetchApiRoute, and decodes the response as JSON
export async function fetchApiJson(
  url: RequestInfo | URL,
  init?: RequestInit
): Promise<any> {
  return await (await fetchApiRoute(url, init)).json();
}

/**
 *  Wrap Elements with QueryClient & DndProvider similar to _app.tsx
 *
 *  Also overides window.fetch to use fetchApiRoute for testing
 * */
export function render(children: JSX.Element | JSX.Element[]): RenderResult {
  window.fetch = fetchApiRoute;
  return testRender(
    <div data-testid="test-root-element">
      <QueryClientProvider client={new QueryClient()}>
        <DndProvider backend={HTML5Backend}>{children}</DndProvider>
      </QueryClientProvider>
    </div>
  );
}

const localStorage: { [key: string]: string | null } = {};

export function getMockStorage(key: string): string | null {
  return localStorage[key];
}

export function setMockStorage(key: string, value: string): void {
  localStorage[key] = value;
}

export function removeMockStorage(key: string): void {
  localStorage[key] = null;
}

export function clearMockStorage(): void {
  for (const k in localStorage) {
    localStorage[k] = null;
  }
}

/// Configure local storage and set pre-existing data in local storage
export function buildLocalStorage(user?: UserMajor): void {
  window.localStorage.getItem = getMockStorage;
  window.localStorage.setItem = setMockStorage;
  window.localStorage.removeItem = removeMockStorage;
  window.localStorage.clear = clearMockStorage;
  if (user !== undefined) {
    window.localStorage.setItem("user_major", JSON.stringify(user));
  }
}

export function mockUseQuery<T>(
  _key: any,
  fetch: () => Promise<T>
): UseQueryResult<T> & { fetch: Promise<T> } {
  return {
    fetch: fetch(),
    data: undefined,
    error: null,
    isError: false,
    isIdle: false,
    isLoading: true,
    isLoadingError: false,
    isRefetchError: false,
    isSuccess: false,
    status: "loading",
    dataUpdatedAt: Date.now(),
    errorUpdatedAt: Date.now(),
    failureCount: 0,
    errorUpdateCount: 0,
    isFetched: true,
    isFetchedAfterMount: true,
    isFetching: false,
    isPlaceholderData: false,
    isPreviousData: false,
    isRefetching: false,
    isStale: false,
    refetch: async (_options: any) => {
      throw new Error("");
    },
    remove: () => { }
  };
}

export function setupStandaloneQuery(): void {
  jest.mock("react-query", () => ({
    useQuery: mockUseQuery
  }));
}

export async function waitForQuery<T>(
  query: () => UseQueryResult<T>
): Promise<T> {
  // @ts-expect-error fetch only exists on our mock
  const temp = await query().fetch;
  return temp;
}

const toBeUnique: MatcherFunction<[]> = (actual) => {
  if (!Array.isArray(actual)) {
    throw new Error("actual is not an object");
  }
  if (actual === null) {
    return {
      message: () => "null is not an array",
      pass: false
    };
  }
  for (const i of actual) {
    const pos = actual.find((val, idx) => idx > i && actual[i] === val);
    if (pos !== undefined) {
      return {
        message: () => `Indicies ${i}, and ${pos} are the same`,
        pass: false
      };
    }
  }
  return {
    message: () => "All elements are unique",
    pass: true
  };
};

const toBeHidden: MatcherFunction<[]> = (actual) => {
  if (!(actual instanceof HTMLElement)) {
    throw new Error("actual is not an HTMLElement");
  }
  let cur = actual;
  console.log("child", cur.ariaHidden);
  while (cur.ariaHidden == null || cur.ariaHidden === "false") {
    if (cur.dataset.testid === "Semester") {
      console.log(cur.ariaHidden);
    }
    const next = cur.parentElement;
    if (next === null) {
      return {
        message: () => "Element (or parent) is not aria-hidden",
        pass: false
      };
    }
    cur = next;
  }
  return {
    message: () => "Element (or parent) is aria-hidden",
    pass: true
  };
};

expect.extend({
  toBeUnique,
  toBeHidden
});

declare module "expect" {
  interface AsymmetricMatchers {
    toBeUnique: () => void;
    toBeHidden: () => void;
  }
  interface Matchers<R> {
    toBeUnique: () => R;
    toBeHidden: () => R;
  }
}

export function courseSemestersCheck(semesters: string): void {
  /*
    Available options: FA,WI,SP,SU

    semesters can start with any of the options
    then be followed by a comma and any of the options

    Note: Data such as FA,FA will match but FA,bad_data will not
  */
  // Expect the format to match
  expect(semesters).toMatch(/^(FA|WI|SP|SU)(,(FA|WI|SP|SU))*$/);
}

export function createMockToken(): void {
  updateClient((c) => {
    c.verifyIdToken = async (options: VerifyIdTokenOptions) => {
      const [test, id] = options.idToken.split(":");
      if (test !== "TEST_TOKEN") {
        throw new Error("Invalid Token");
      }
      return new LoginTicket("", {
        sub: id,
        aud: c._clientId ?? "",
        exp: Date.now() + 1000000,
        iat: Date.now() - 100,
        iss: "https://accounts.google.com",
        at_hash: ""
      });
    };
  });
}

export function mockToken(id: string): string {
  return `TEST_TOKEN:${id}`;
}

export function mockUserInfo(id: string): User {
  return {
    cred: mockToken(id),
    info: {
      sub: id,
      picture: "",
      name: id,
      email: `${id}@gmail.com`
    }
  };
}

const db = {
  db: new PromisedDatabase(),
  opened: false
};

export async function setupMockUserDB(): Promise<void> {
  createMockToken();
  setUserDb(db.db);
  if (!db.opened) {
    db.opened = true;
    await db.db.open(":memory:");
    await db.db.run(
      "CREATE TABLE `user` ( `idUser` TEXT PRIMARY KEY, `role` TEXT NOT NULL)"
    );
    await db.db.run(
      "CREATE TABLE `schedule` ( `userID` TEXT, `name` TEXT, `timestamp` DATE DEFAULT (datetime('now','localtime')), `scheduleData` TEXT NOT NULL, PRIMARY KEY(`userID`, `name`))"
    );
  }
}
