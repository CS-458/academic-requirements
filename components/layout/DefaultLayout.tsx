import {
  AppBar,
  Box,
  Toolbar,
  Typography,
  Button,
  Popover,
  MenuList,
  MenuItem,
  Tooltip
} from "@mui/material";

import MenuDrawer from "../NavigationMenu";
import {
  CredentialResponse,
  GoogleLogin,
  googleLogout
} from "@react-oauth/google";
import { useEffect, useState } from "react";
import jwtDecode from "jwt-decode";
import { User, UserLogin, UserInfo } from "../../services/user";
import LogoLink from "./LogoLink";
import Router from "next/router";

let curTimeout: undefined | NodeJS.Timeout;

export default function DefaultLayout(props: {
  children: JSX.Element | JSX.Element[];
}): JSX.Element {
  const [user, setUser] = useState<User | undefined>(undefined);
  function setUserAndTimeout(user: User | undefined): void {
    setUser(user);
    clearTimeout(curTimeout);
    if (user !== undefined) {
      // settimeout for expiry
      curTimeout = setTimeout(() => {
        setUser(undefined);
      }, user.info.exp * 1000 - Date.now());
    }
  }

  const [firstLoad, setFirstLoad] = useState(true);
  useEffect(() => {
    const loaded = localStorage.getItem("google-login");
    if (loaded !== null) {
      const user: User = JSON.parse(loaded);
      const now = Date.now() / 1000;
      if (user.info.exp >= now && user.info.nbf <= now) {
        setUserAndTimeout(user);
      }
    }
    setFirstLoad(false);
  }, []);

  function logout(): void {
    localStorage.removeItem("google-login");
    setUserAndTimeout(undefined);
    // Just in case
    googleLogout();
  }

  function responseMessage(token: CredentialResponse): void {
    if (token.credential !== undefined) {
      const jwt: UserInfo = jwtDecode(token.credential);
      const user = { info: jwt, cred: token.credential };
      setUserAndTimeout(user);
      localStorage.setItem("google-login", JSON.stringify(user));
    }
  }
  function errorMessage(): void {
    console.error("Login failed");
  }
  const [picture, setPicture] = useState(user?.info.picture);
  useEffect(() => {
    setPicture(user?.info.picture);
  }, [user]);

  function UserLoginElement(): JSX.Element {
    // First load is true until we have attempted to load a user from localStorage.
    const [anchorEl, setAnchorEl] = useState<HTMLButtonElement | null>(null);

    const handleClick = (event: React.MouseEvent<HTMLButtonElement>): void => {
      setAnchorEl(event.currentTarget);
    };

    const handleClose = (): void => {
      setAnchorEl(null);
    };

    if (firstLoad) {
      return <></>;
    }
    if (user === undefined) {
      return (
        <div data-testid="google-login-button">
          <GoogleLogin
            onSuccess={responseMessage}
            onError={errorMessage}
            useOneTap
          />
        </div>
      );
    }

    const open = anchorEl !== null;
    const id = open ? "simple-popover" : undefined;
    function handleListKeyDown(event: React.KeyboardEvent): void {
      if (event.key === "Tab") {
        event.preventDefault();
        setAnchorEl(null);
      } else if (event.key === "Escape") {
        setAnchorEl(null);
      }
    }
    function toAccountPage(): void {
      Router.push("/account").catch(console.error);
    }

    return (
      <Typography variant="h5" component="div" sx={{ fontSize: "1rem" }}>
        <Button aria-describedby={id} onClick={handleClick}>
          <Tooltip title={"Manage Profile"}>
            <img
              src={picture}
              onError={(e) => {
                console.log(e);
                const target = e.target;
                if (target instanceof HTMLImageElement) {
                  target.src = "/defaultProfile.png";
                }
              }}
              style={{ height: "2em", borderRadius: "50%" }}
              data-testid="account-picture"
            />
          </Tooltip>
        </Button>
        <Popover
          id={id}
          anchorEl={anchorEl}
          open={open}
          onClose={handleClose}
          anchorOrigin={{
            vertical: "bottom",
            horizontal: "right"
          }}
          transformOrigin={{
            vertical: "top",
            horizontal: "right"
          }}
        >
          <MenuList
            autoFocusItem={open}
            id="composition-menu"
            aria-labelledby="composition-button"
            onKeyDown={handleListKeyDown}
          >
            <MenuItem onClick={toAccountPage}>Account</MenuItem>
            <MenuItem onClick={logout}>Logout</MenuItem>
          </MenuList>
        </Popover>
      </Typography>
    );
  }

  return (
    <div
      style={{
        height: "100vh",
        width: "100vw",
        display: "grid",
        gridTemplateRows: "min-content minmax(0px, auto)"
      }}
    >
      <UserLogin.Provider value={user}>
        <Box sx={{ flexGrow: 1 }}>
          <AppBar
            position="fixed"
            sx={{
              zIndex: (theme) => theme.zIndex.drawer + 1,
              position: "relative"
            }}
          >
            <Toolbar>
              <MenuDrawer />
              <LogoLink />
              <Typography variant="h5" component="div" sx={{ flexGrow: 1 }} />
              <UserLoginElement />
            </Toolbar>
          </AppBar>
        </Box>
        {props.children}
      </UserLogin.Provider>
    </div>
  );
}
