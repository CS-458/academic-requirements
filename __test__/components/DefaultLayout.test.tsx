import "@testing-library/jest-dom";
// import { render, screen } from "@testing-library/react";
import { screen, waitFor } from "@testing-library/react";
import { mockUserInfo, render, setupUser, userId } from "../util";
import DefaultLayout from "../../components/layout/DefaultLayout";
import { act } from "react-test-renderer";

test("Render Default Layout", async () => {
  // const user = setupUser();
  localStorage.removeItem("google-login");
  const page = render(
    <DefaultLayout>
      <div />
    </DefaultLayout>
  );
  expect(screen.getByTestId("google-login-button")).toBeInTheDocument();
  expect(page.baseElement).toMatchSnapshot();
});

test("Render Default Layout w/user", async () => {
  const user = setupUser();
  localStorage.setItem("google-login", JSON.stringify(mockUserInfo(userId())));

  const page = render(
    <DefaultLayout>
      <div />
    </DefaultLayout>
  );

  await user.click(screen.getByTestId("account-picture"));

  expect(screen.getByText(/Logout/i)).toBeInTheDocument();
  expect(page.baseElement).toMatchSnapshot();

  await user.click(screen.getByText(/Logout/i));

  expect(screen.getByTestId("google-login-button")).toBeInTheDocument();
});

test("Render Default Layout w/expired user", async () => {
  const token = mockUserInfo(userId());
  token.info.exp = Date.now() / 1000 - 10;
  localStorage.setItem("google-login", JSON.stringify(token));

  render(
    <DefaultLayout>
      <div />
    </DefaultLayout>
  );

  expect(screen.getByTestId("google-login-button")).toBeInTheDocument();
});

test("Render Default Layout w/expiring user", async () => {
  const token = mockUserInfo(userId());
  // Default timeout is 5000, and we need to wait at least 1000. We don't do anything else here, so it isn't an issue.
  token.info.exp = Date.now() / 1000 + 2;
  localStorage.setItem("google-login", JSON.stringify(token));

  const user = setupUser();
  const page = render(
    <DefaultLayout>
      <div />
    </DefaultLayout>
  );

  await user.click(screen.getByTestId("account-picture"));
  expect(screen.getByText(/Logout/i)).toBeInTheDocument();
  expect(page.baseElement).toMatchSnapshot();

  await waitFor(
    () => expect(screen.getByTestId("google-login-button")).toBeInTheDocument(),
    { timeout: 3000 }
  );
});

test("Render Default Layout w/error for picture", async () => {
  const token = mockUserInfo(userId());
  // Default timeout is 5000, and we need to wait at least 1000. We don't do anything else here, so it isn't an issue.
  localStorage.setItem("google-login", JSON.stringify(token));

  const user = setupUser();
  const page = render(
    <DefaultLayout>
      <div />
    </DefaultLayout>
  );

  const image = screen.getByTestId("account-picture") as HTMLImageElement;

  await user.click(image);
  expect(screen.getByText(/Logout/i)).toBeInTheDocument();
  await act(() => {
    image.dispatchEvent(new UIEvent("error"));
  });
  expect(image.src.endsWith("/defaultProfile.png")).toBe(true);
});
