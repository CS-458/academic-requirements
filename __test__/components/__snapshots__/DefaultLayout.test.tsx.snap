// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Default Layout 1`] = `
<body>
  <div>
    <div
      data-testid="test-root-element"
    >
      <div
        style="height: 100vh; width: 100vw; display: grid; grid-template-rows: min-content minmax(0px, auto);"
      >
        <div
          class="MuiBox-root css-i9gxme"
        >
          <header
            class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-i1wygj-MuiPaper-root-MuiAppBar-root"
          >
            <div
              class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
            >
              <div>
                <button
                  aria-label="menu"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeMedium css-134qg7o-MuiButtonBase-root-MuiIconButton-root"
                  data-testid="menu"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="MenuIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
              <a
                data-testid="stout-logo-link"
                href="https://www.uwstout.edu/"
                target="_blank"
              >
                <img
                  alt="logo"
                  height="60"
                  src="/logo-new.svg"
                />
              </a>
              <div
                class="MuiTypography-root MuiTypography-h5 css-1f8a19r-MuiTypography-root"
              />
              <div
                data-testid="google-login-button"
              >
                <div
                  style="height: 40px;"
                />
              </div>
            </div>
          </header>
        </div>
        <div />
      </div>
    </div>
  </div>
  <script
    defer=""
    src="https://accounts.google.com/gsi/client"
  />
</body>
`;

exports[`Render Default Layout w/expiring user 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div
      data-testid="test-root-element"
    >
      <div
        style="height: 100vh; width: 100vw; display: grid; grid-template-rows: min-content minmax(0px, auto);"
      >
        <div
          class="MuiBox-root css-i9gxme"
        >
          <header
            class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-i1wygj-MuiPaper-root-MuiAppBar-root"
            style="padding-right: 1024px;"
          >
            <div
              class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
            >
              <div>
                <button
                  aria-label="menu"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeMedium css-134qg7o-MuiButtonBase-root-MuiIconButton-root"
                  data-testid="menu"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="MenuIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
              <a
                data-testid="stout-logo-link"
                href="https://www.uwstout.edu/"
                target="_blank"
              >
                <img
                  alt="logo"
                  height="60"
                  src="/logo-new.svg"
                />
              </a>
              <div
                class="MuiTypography-root MuiTypography-h5 css-1f8a19r-MuiTypography-root"
              />
              <div
                class="MuiTypography-root MuiTypography-h5 css-1r5njk-MuiTypography-root"
              >
                <button
                  aria-describedby="simple-popover"
                  class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
                  tabindex="0"
                  type="button"
                >
                  <img
                    data-testid="account-picture"
                    src=""
                    style="height: 2em; border-radius: 50%;"
                  />
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  >
                    <span
                      class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
                      style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
                    >
                      <span
                        class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
                      />
                    </span>
                  </span>
                </button>
              </div>
            </div>
          </header>
        </div>
        <div />
      </div>
    </div>
  </div>
  <script
    defer=""
    src="https://accounts.google.com/gsi/client"
  />
  <div
    class="MuiPopover-root MuiModal-root css-1i0e6f3-MuiModal-root-MuiPopover-root"
    id="simple-popover"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiBackdrop-invisible MuiModal-backdrop css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiPopover-paper css-3bmhjh-MuiPaper-root-MuiPopover-paper"
      style="opacity: 1; transform: none; transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
      tabindex="-1"
    >
      <ul
        aria-labelledby="composition-button"
        class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
        id="composition-menu"
        role="menu"
        tabindex="-1"
      >
        <li
          class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
          role="menuitem"
          tabindex="0"
        >
          Account
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </li>
        <li
          class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
          role="menuitem"
          tabindex="-1"
        >
          Logout
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </li>
      </ul>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;

exports[`Render Default Layout w/user 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div
      data-testid="test-root-element"
    >
      <div
        style="height: 100vh; width: 100vw; display: grid; grid-template-rows: min-content minmax(0px, auto);"
      >
        <div
          class="MuiBox-root css-i9gxme"
        >
          <header
            class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-i1wygj-MuiPaper-root-MuiAppBar-root"
            style="padding-right: 1024px;"
          >
            <div
              class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
            >
              <div>
                <button
                  aria-label="menu"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeMedium css-134qg7o-MuiButtonBase-root-MuiIconButton-root"
                  data-testid="menu"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="MenuIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
              <a
                data-testid="stout-logo-link"
                href="https://www.uwstout.edu/"
                target="_blank"
              >
                <img
                  alt="logo"
                  height="60"
                  src="/logo-new.svg"
                />
              </a>
              <div
                class="MuiTypography-root MuiTypography-h5 css-1f8a19r-MuiTypography-root"
              />
              <div
                class="MuiTypography-root MuiTypography-h5 css-1r5njk-MuiTypography-root"
              >
                <button
                  aria-describedby="simple-popover"
                  class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
                  tabindex="0"
                  type="button"
                >
                  <img
                    data-testid="account-picture"
                    src=""
                    style="height: 2em; border-radius: 50%;"
                  />
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  >
                    <span
                      class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
                      style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
                    >
                      <span
                        class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
                      />
                    </span>
                  </span>
                </button>
              </div>
            </div>
          </header>
        </div>
        <div />
      </div>
    </div>
  </div>
  <script
    defer=""
    src="https://accounts.google.com/gsi/client"
  />
  <div
    class="MuiPopover-root MuiModal-root css-1i0e6f3-MuiModal-root-MuiPopover-root"
    id="simple-popover"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiBackdrop-invisible MuiModal-backdrop css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiPopover-paper css-3bmhjh-MuiPaper-root-MuiPopover-paper"
      style="opacity: 1; transform: none; transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
      tabindex="-1"
    >
      <ul
        aria-labelledby="composition-button"
        class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
        id="composition-menu"
        role="menu"
        tabindex="-1"
      >
        <li
          class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
          role="menuitem"
          tabindex="0"
        >
          Account
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </li>
        <li
          class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
          role="menuitem"
          tabindex="-1"
        >
          Logout
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </li>
      </ul>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
